ApiReview/API.Polly.Core/NoDocs/Polly/ResilienceContext.cs

```diff
@@ -0,0 +1,27 @@
+// Assembly 'Polly.Core'
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
+using System.Runtime.CompilerServices;
+using System.Threading;
+using Polly.Telemetry;
+using Polly.Utils;
+
+namespace Polly;
+
+public sealed class ResilienceContext
+{
+    public string? OperationKey { get; }
+    public CancellationToken CancellationToken { get; set; }
+    public bool IsSynchronous { get; }
```

Comment:
do we need this? 
as discussed last time, with this property in place every async strategy provides different implementations for the different modes within a single method - the analyzers presumable will flag all sync implementations, the contract becomes quite fuzzy because as a strategy author I mightn't even realize that there must be two impls. etc

Reply:
Yes, it's required. The user-delegates might behave differently for sync vs async executions. 

See:
https://github.com/App-vNext/Polly/tree/main/src/Polly.Core#registering-custom-callbacks

Reply:
you missed the point - as discussed last time, I believe we should have two different interfaces and/or methods on the same interface for sync and async flows

Reply:
My point still stands, even if we would have 2 different methods to implement (`ExecuteAsync`, `ExecuteSync`) the user-delegate is the same for both flows.

Reply:
Leaving my comments here since I'm currently developing the chaos strategies as part of the Simmy migration: I think it is practical to have just one method to implement (`ExecuteCoreAsync`) since as a strategy author you don't have duplicate code, so it makes it easier and convenient, on the other hand, it's a bit confusing since you might end up thinking that the strategy is only for async executions unless you dig into the code, then you realize that internally both sync and async executions always call the `ExecuteCoreAsync` implementation. (`Execute` => `ExecuteCoreSync` =>`ExecuteCoreAsync` and `ExecuteAsync` => `ExecuteCoreAsync`). I'd suggest dropping the `Async` suffix and calling it just `ExecuteCore` or maybe using another name like `Run` or something like that. 

Generally speaking, most of the time I lean towards being explicit to avoid confusion and ambiguity, so having two different methods to implement (`ExecuteAsync` and `ExecuteSync`) helps on that matter, however, the trade-off would be that the author needs to implement two methods and most likely "duplicate" code as the old Polly implementation where there were two engine implementations, one for sync and another for async executions (and the code was practically the same).

Said that, personally, I like the current approach I'd just make it clear in the docs that the `ExecuteCoreAsync` method executes safely both flows so that it is clear for the strategy author. On the other hand the static analyzers it's a great point 🤔

Reply:
> however, the trade-off would be that the author needs to implement two methods and most likely "duplicate" code as the old Polly implementation where there were two engine implementations, one for sync and another for async executions (and the code was practically the same).

let's unpack this
we seem to practically have 3 cases to support:
1 - where the actual logic of a strategy can have both sync and async implementation - I believe it would be better to have two methods split upfront so that (again) compiler and analyzers could force the correct usage of the APIs.
2 - where the actual logic of a strategy is sync - in order to avoid duplicating the code, the library can offer an abstract class with the sync signatures and default impl of the async piece
3 - where the actual logic of a strategy is async only - the library can offer an abstract class with the async signatures only and throw an exception if the sync path is used.

now the implementer can decide to either implement (both async and sync) or (just one and let the library connect the dots) fully benefiting from the compile time help

Reply:
From the 10+ strategies that we've implemented so far here is the summary for your use cases:

- 1: Utilizes retry and hedging. Even within this, we have only 2-3 lines of code enclosed within an `if` statement. 
- 2: Currently has no applicable use case.
- 3: Also lacks a relevant use case.

Most custom strategies merely conduct actions before and after the callback. The split, it could be argued, is artificially induced. Refer to these examples:

- [RequestMessageSnapshotStrategy.cs](https://github.com/dotnet/extensions/blob/519cfb2ff3f8e09dc6e2090a34896716ee58609b/src/Libraries/Microsoft.Extensions.Http.Resilience/Hedging/Internals/RequestMessageSnapshotStrategy.cs)
- [RoutingResilienceStrategy.cs](https://github.com/dotnet/extensions/blob/519cfb2ff3f8e09dc6e2090a34896716ee58609b/src/Libraries/Microsoft.Extensions.Http.Resilience/Routing/Internal/RoutingResilienceStrategy.cs)

Even for use-case 1, the implementation remains identical for both asynchronous and synchronous paths across all strategies. Polly V8's design assures this by passing the synchronous callback for synchronous executions to the strategy.

- [RetryResilienceStrategy.cs](https://github.com/App-vNext/Polly/blob/main/src/Polly.Core/Retry/RetryResilienceStrategy.cs)
- [HedgingResilienceStrategy.cs](https://github.com/App-vNext/Polly/blob/main/src/Polly.Core/Hedging/HedgingResilienceStrategy.cs)

In an effort to explore the impact of having separate `Sync` and `Async` paths, I attempted to implement this split as part of an exercise. However, I abandoned it midway when it became apparent that it would only serve to complicate the APIs.

In summary, the current API allows for almost "free" support for synchronous executions, while still remaining highly optimized for asynchronous scenarios.

Reply:
so, you're saying that we have only the cases where the logic must be separated into sync and async flows, then I don't understand why merged the two methods

Reply:
> so, you're saying that we have only the cases where the logic must be separated into sync and async flows, then I don't understand why merged the two methods.

Where did I said that? Quote from above:

> Even for use-case 1, the implementation remains identical for both asynchronous and synchronous paths across all strategies. Polly V8's design assures this by passing the synchronous callback for synchronous executions to the strategy. 

What this means is that there is no need to separate the logic into synchronous and asynchronous flows. The current implementation handles both flows in a unified manner, without requiring any division. The implementation is optimized to cater to both synchronous and asynchronous scenarios using a single code path with minimal branching. 

I hope this clarifies my point.

cc: @vany0114 just in case there is anything left to add

Reply:
> Even for use-case 1, the implementation remains identical for both asynchronous and synchronous paths across all strategies.

if so, then it probably falls into the second group because by definition I gave strategies in the first group have different implementations

Reply:
I have created a draft PR to demonstrate how the API would look + the changes necessary:

#1434

Reply:
I am posting here a summary of the offline discussion we had with @martintmk.
The problems with the current implementation are:
1. the current ValueTask-based approach hinders the ability of the compiler to validate the sync implementation of policies. Essentially, all the policies are async and the policy authors (maintainers included) need to manually and carefully study their code to prevent their implementations from starting async tasks.
2. this actually led me to a question of what exact promise the library gives in regard to sync executions. I see only two options: (1) when a policy is executed synchronously, the call blocks the thread until the invocation is finished and all operations invoked by the policy internally are executed within the same thread, including all the callbacks, etc., (2) when a policy is executed synchronously, the call blocks the thread until the invocation is finished, but the dependent operations might be executed in other threads. in case the library implements the second contract, can't the sync method be implemented as an extension `ExecuteAsync.GetAwaiter().GetResult()`  and if so, why do we need the flag in the first place? What could motivate a policy author to have a sync-path in the policy code?

Reply:
Fixed in #1494